File: slimjim49j/day_1/c/day_1.cpp
#include <stdio.h>
#include <Windows.h>

typedef char s8;

typedef unsigned char u8;
typedef unsigned short u16;
typedef unsigned int u32;
typedef unsigned long u64;

struct OpenFileReturn
{
    u64 bytesRead;
    void *file;
};

OpenFileReturn openFile(char *fileName)
{
    HANDLE fileHandle = CreateFileA(
        fileName,
        GENERIC_READ,
        FILE_SHARE_READ,
        0,
        OPEN_EXISTING,
        0,
        0);
    OpenFileReturn result = {};

    if (fileHandle)
    {
        LARGE_INTEGER fileSize;
        if (GetFileSizeEx(fileHandle, &fileSize))
        {
            void *memory = VirtualAlloc(0, fileSize.QuadPart, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
            if (memory)
            {

                DWORD bytesRead;
                // TODO assert, readfile doesn't take a 64 bit val
                if (ReadFile(
                        fileHandle,
                        memory,
                        fileSize.QuadPart,
                        &bytesRead,
                        0) &&
                    fileSize.QuadPart == bytesRead)
                {
                    result.bytesRead = bytesRead;
                    result.file = memory;
                }
                else
                {
                    VirtualFree(memory, 0, MEM_RELEASE);
                }
            }
            else
            {
                return OpenFileReturn{};
            }
        }
        else
        {
            return OpenFileReturn{};
        }
    }
    else
    {
        return OpenFileReturn{};
    }

    CloseHandle(fileHandle);
    return result;
}

void main()
{
    const char *digits[10] = {};
    digits[0] = "zero";
    digits[1] = "one";
    digits[2] = "two";
    digits[3] = "three";
    digits[4] = "four";
    digits[5] = "five";
    digits[6] = "six";
    digits[7] = "seven";
    digits[8] = "eight";
    digits[9] = "nine";

    OpenFileReturn file = openFile("input");

    if (file.bytesRead && file.file) {
        u8 * input = (u8 *)file.file;
        
    }

    return;
}

File: slimjim49j/day_1/go/day_1.go
package main

import (
	"fmt"
	"os"
)

var digits = [10]string{"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	input, err := os.ReadFile("input")
	check(err)

	sum := 0
	seenFirst := false
	first := 0
	last := 0

	for i := 0; i < len(input); i++ {
		ch := input[i]
		if ch >= 48 && ch <= 57 {
			n := ch - 48
			if !seenFirst {
				first = int(n)
				seenFirst = true
			}
			last = int(n)
		} else if ch >= 97 && ch <= 122 {
			for j := 0; j < len(digits); j++ {
				digit := digits[j]
				isMatch := true
				for k := 0; k < len(digit); k++ {
					srcCh := input[i+k]
					cmpCh := digit[k]
					if srcCh != cmpCh {
						isMatch = false
						break
					}
				}
				if isMatch {
					if !seenFirst {
						first = j
						seenFirst = true
					}
					last = j
					break
				}
			}
		} else if ch == 10 {
			sum += (first * 10) + last
			seenFirst = false
			first = 0
			last = 0
		}
	}

	fmt.Printf("%d\n", sum)
}
